src/main/java/appeng/api/storage/IMEInventory.java

```diff
@@ -23,24 +23,28 @@
 
 package appeng.api.storage;
 
+import java.util.Objects;
+
+import com.google.common.base.Preconditions;
+
 import net.minecraft.world.item.ItemStack;
 
 import appeng.api.config.Actionable;
 import appeng.api.networking.security.IActionSource;
-import appeng.api.storage.data.IAEStack;
-import appeng.api.storage.data.IAEStackList;
+import appeng.api.storage.data.AEKey;
+import appeng.api.storage.data.KeyCounter;
 
 /**
  * AE's Equivalent to IInventory, used to reading contents, and manipulating contents of ME Inventories.
- *
+ * <p>
  * Implementations should COMPLETELY ignore stack size limits from an external view point, Meaning that you can inject
  * Integer.MAX_VALUE items and it should work as defined, or be able to extract Integer.MAX_VALUE and have it work as
  * defined, Translations to MC's max stack size are external to the AE API.
  * <p/>
  * If you want to request at most a stack of an item, you need to use {@link ItemStack#getMaxStackSize()} before
  * extracting from this inventory.
  */
-public interface IMEInventory<T extends IAEStack> {
+public interface IMEInventory<T extends AEKey> {
```

Comment:
I'd remove the default from `insert`, `extract` and `getAvailableStacks` as it can be quite error-prone I think.